walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,118 @@
 // limitations under the License.
 
 package com.google.sps;
-
+// import com.google.sps.servlets.TimeRange;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> ret = new ArrayList<TimeRange>();
+
+    Set<String> mandatoryAttendees =  new HashSet<String>();
+    mandatoryAttendees.addAll(request.getAttendees()); 
+
+    Set<String> optionalAttendees = new HashSet<String>();
+    optionalAttendees.addAll(request.getOptionalAttendees());
+
+    Set<String> allAttendees = new HashSet<String>();
+    allAttendees.addAll(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+
+    long duration = request.getDuration();
+
+    // no options for too long of a request
+    if (duration > 24*60){
+      return ret;
+    }
+
+    // options for no attendees
+    if (allAttendees.size() == 0) {
+      ret.add(TimeRange.WHOLE_DAY);
+      return ret;
+    }
+
+    // find free times
+    List<TimeRange> freeTimes = new ArrayList<TimeRange>(findFreeTimes(allAttendees, events, request));
+    if (freeTimes.isEmpty()) {
+      // ignore optional attendees if there are no free times across both mandatory and optional attendees
+      freeTimes = new ArrayList<TimeRange>(findFreeTimes(mandatoryAttendees, events, request));
+    } else {
+      return freeTimes;
+    }
+
+    // adhere to additional test requirement
+    if (mandatoryAttendees.isEmpty()) {
+      return new ArrayList<TimeRange>();
+    }
+    return freeTimes;
+  }
+
+  private List<TimeRange> mergeOverlaps(List<TimeRange> blocked) {
```

Comment:
this doesn't seem to be used. do we need it?

Reply:
forgot to delete from older code